### ISSUE TITLE
Each start: "Duplicate dictionary ID found" & affected indices rebuilt

### ISSUE THREAD

**Issue by twwn** [xiaoyifang/goldendict-ng#1082]
**Describe the bug**
Upon each start, Goldendict-ng writes several `"Duplicate dictionary ID found: ID={the_id},` warnings and will expensively rebuild several indices (of the affected dictionaries, it seems), accompanied by `DatabaseLockError: Unable to get write lock on /home/.cache/goldendict/index/{the_id}: already locked`.

This affects several EPWING dictionaries. For some the warning is given once, others twice or trice.

Please see the attached txt for an example log: [console.txt](https://github.com/xiaoyifang/goldendict-ng/files/12432498/console.txt). I quit the program after a while as a full rebuilding only heats the room.

**To Reproduce**
1. Add dictionaries to Goldendict
2. Restart

**Affected Dictionaries**

In attached txt: [console.txt](https://github.com/xiaoyifang/goldendict-ng/files/12432498/console.txt)

**Expected behavior**
Each index is built only once.

**Screenshots**
n/a

**OS and software versions**

```
Goldendict-ng 23.07.25.a200989d at 2023-08-23T11:27:22
Qt 6.5.2 GCC 13.2.1 20230801 arch linux 6.4.12-zen1-1-zen x86_64-little_endian-lp64
Flags:USE_XAPIAN MAKE_ZIM_SUPPORT USE_ICONV MAKE_CHINESE_CONVERSION_SUPPORT
```

**Additional context**
No change after deleting the indices folder entirely and restarting GD-ng after a full rebuild.

This isn't the only EPWING problem: [This Goldendict bug](https://github.com/goldendict/goldendict/issues/948) also persists unchanged.

When clicking "Copy dictionaries list", some are listed four times in a row.

**Comment by xiaoyifang** [comment]
> "Duplicate dictionary ID found: ID={the_id},

This is not an error , this is usually happens when you have added the same dictionary folder twice ,  such as folder  `folder/a/b/c`

you added `folder/a/b` then add `folder/a/b/c`   which will give a duplicateId warning.

> DatabaseLockError: Unable to get write lock on /home/.cache/goldendict/index/{the_id}: already locked.

this should be caused by the duplicateId issue.  it seems will cause no harm.

**Comment by xiaoyifang** [comment]
> This isn't the only EPWING problem: [This Goldendict bug](https://github.com/goldendict/goldendict/issues/948) also persists unchanged.

Can you give a more detailed steps to reproduce this?   
1 , the dictionary ,(upload here if necessary).
2,  the searched terms to reproduce this

**Comment by twwn** [comment]
> This is not an error , this is usually happens when you have added the same dictionary folder twice , such as folder `folder/a/b/c`
> 
> you added `folder/a/b` then add `folder/a/b/c` which will give a duplicateId warning.

I'm not sure I understand correctly? They've been added only once under _Sources_ → _Files_ (parent folder, then "recursive" enabled) and their files are not duplicated. Clicking "rescan" does cause GD-ng to print the `Duplicate dictionary ID` messages, though.

Some their folders have subfolders but those carry separate content and don't match the number of times GD-ng reports them as repeated. These settings are inherited from the old GD, which did not reindex.

>1 , the dictionary ,(upload here if necessary).

>ビジネス技術実用英語大辞典第４版
>岩波日本史辞典ＣＤ‐ＲＯＭ版
>広辞苑第六版

They're too large (広辞苑 alone is 3.5GB), but you can google "Koujien" together with "cornucopia" to find it. EBView works fine with the same files.

Every single term is affected. The search completion lists some correct suggestions, but not all (e.g., 赤ん should suggest 赤ん坊, but doesn't suggest anything in GD-ng).

**Comment by xiaoyifang** [comment]
> Every single term is affected. The search completion lists some correct suggestions, but not all (e.g., 赤ん should suggest 赤ん坊, but doesn't suggest anything in GD-ng).

have you tried to rebuild the index of a certain dictionary ?

**Comment by twwn** [comment]
It got wiped when I deleted the entire index folder, yes.

The index isn't the real issue though, it's that no entry actually loads. There's always some EPWING error and/or occasionally a wrong entry:

Suggestion "赤" yields…

>Article reading error: EB library report error EB "eb_read_heading" function error: EB_ERR_UNEXP_TEXT (unexpected format in a text file) on page 9122, offset 1932

… followed by this weird unrelated entry

>御門　享保　　３　戊戌
>1719　中御門　享保　　４　己亥
>[followed by 16 more such lines]

Suggestion "あか 赤" yields just the same unrelated entry without the explicit function error.

Suggestion "赤外" yields

>Article reading error: EB library report error EB "eb_seek_text" function error: EB_ERR_NO_TEXT (no text file) on page 13921, offset 748

Entry "あかし 赤し紅し" yields some unrelated garbled introductory text without function error, beginning with

>籬�米鋲俶阡闢蓖凱限浜限浜限浜限浜限浜限浜限浜限浜限浜鍋跂鱸�

The other affected dictionaries likewise yield either "eb_seek_text" or "eb_read_heading" errors, except I've seen GD-ng output two errors for one entry rather than just one.

I don't know what libebu functions GD calls, but seems like something on my system causes it to do so with wrong parameters. (I have four UTF-8 locales enabled in my `/etc/locale.gen`, in case that could matter.)

**Comment by xiaoyifang** [comment]
> "あか 赤"
>"あかし こくさい" 
>"あかし 赤し紅し"

all contain space in them ,does this normal?

I tried to search with ebwin ,can not found these entries neigher.
![image](https://github.com/xiaoyifang/goldendict-ng/assets/105986/418a2712-5fde-4b28-969f-8dacd64be77f)

**Comment by xiaoyifang** [comment]
epwing is not a dictionary format ,
I guess the problemed dictionaries has some informal headwords .
One way to bypass this is search with fulltext

![image](https://github.com/xiaoyifang/goldendict-ng/assets/105986/ae17df95-3814-4741-8fc6-8c169dd8a21d)

**Comment by xiaoyifang** [comment]
managed to reproduce this on ubuntu,

seems more likely a libeb issue.  

The dictionary work well on Windows, which means the same code goes well on Windows but not on Linux, the difference is the libeb library.

**Comment by xiaoyifang** [comment]
![图片](https://github.com/xiaoyifang/goldendict-ng/assets/105986/cfa72136-759d-41b7-b456-d4cd1f16958c)
![图片](https://github.com/xiaoyifang/goldendict-ng/assets/105986/a3e086a0-47bc-47df-a48a-88de3e5b8e89)
![图片](https://github.com/xiaoyifang/goldendict-ng/assets/105986/366bb550-a34e-4232-968e-054196c444f9)

**Comment by xiaoyifang** [comment]
> `"Duplicate dictionary ID found: ID={the_id},`

This has been fixed too.

**Comment by twwn** [comment]
Thank you! The entries do load properly now.

Looking at the console output of 78ef410, I see a few more remaining issues:

1. GD-ng seems to search for non-existent EPWING folders. Among others, that yields
```
Epwing: Building the index for dictionary in directory /dict/Eiwaei senmon/Rikagaku/about [does not exist]
"Epwing dictionary initializing failed: /dict/Eiwaei senmon/Rikagaku/about, error: EB library report error EB \"eb_text\" function error: EB_ERR_NO_SUCH_SEARCH (no such search method)\n"
```

2. There also some failures with existing folders (which GD reports as lower-case when they are actually in caps) that cause other errors:
```
Epwing: Building the index for dictionary in directory /dict/Kokugo_misc/Tegami/kokugo [exists, but as KOKUGO]
"Epwing dictionary initializing failed: /dict/Kokugo_misc/Tegami/kokugo, error: EB library report error EB \"eb_forward_text\" function error: EB_ERR_UNEXP_TEXT (unexpected format in a text file) on page 14979, offset 1626\n"

Epwing: Building the index for dictionary in directory /dict/Kokugo_senmon/Jiyuhouritu/houritu [exists, but as HOURITU]
"Epwing dictionary initializing failed: /dict/Kokugo_senmon/Jiyuhouritu/houritu, error: EB library report error EB \"eb_forward_text\" function error: EB_ERR_FAIL_READ_TEXT (failed to read a text file) on page 1737, offset 4\n"

Epwing: Building the index for dictionary in directory /dict/Kokugo_senmon/Nihonsi/furoku [exists, but as FUROKU]
"Epwing dictionary initializing failed: /dict/Kokugo_senmon/Nihonsi/furoku, error: EB library report error EB \"forwardText\" function error: EB_ERR_END_OF_CONTENT (end of content) on page 11, offset 2002\n"

Epwing: Building the index for dictionary in directory /dict/Kokugo_senmon/Sejinmei/furoku [exists, but as FUROKU]
"Epwing dictionary initializing failed: /dict/Kokugo_senmon/Sejinmei/furoku, error: EB library report error EB \"forwardText\" function error: EB_ERR_END_OF_CONTENT (end of content) on page 11, offset 1796\n"
```

The case mismatch doesn't necessarily seem to be the issue, though: An EPWING copy of **JM**dict (same capitalization for its folder) has GD-ng report no error despite it searching for "**jm**dict".

These errors repeat at each start of GD-ng.

3. Some dictionaries like 世界大百科事典 cause GD-ng to print a lot of "dehsinif" lines during the full-text index building. I guess that's not an actual error, though.

4. Some dictionaries cause *"Skipped too long headword"* warnings, eg. Ruigigo with one like
`Skipped too long headword:  "   見つかるみつかる探るさくるあさる尋ねるたすねる求めるも" size: 160`

5. Compared to ebview, GD-ng returns a fraction of the search results. For instance, searching 広辞苑 for 赤 causes ebview to return several hundred entries with that character in their headword, whereas GD-ng returns about 40. (ebview has several somewhat weird [search methods](https://ebview.sourceforge.net/help/en/index.html#SEC3), but this also occurs with its restrictive "Forward Search" one, in which only matches entries that have the query at their headword's beginning.)

**Comment by xiaoyifang** [comment]
> 1. GD-ng seems to search for non-existent EPWING folders. Among others, that yields

should be fixed with https://github.com/xiaoyifang/goldendict-ng/pull/1092

> There also some failures with existing folders (which GD reports as lower-case when they are actually in caps) that cause other errors:

can you upload one of the dictionary for debug



> 3\. Some dictionaries like 世界大百科事典 cause GD-ng to print a lot of "dehsinif" lines during the full-text index building. I guess that's not an actual error, though.

no error.  normal check.





> 4\. Some dictionaries cause _"Skipped too long headword"_ warnings, for instance Ruigigo with several like
>     `Skipped too long headword:  "   見つかるみつかる探るさくるあさる尋ねるたすねる求めるも" size: 160`

This is designed .  Epwing is not a dictionary format,but an ebook format ,  some headword is indeed too long.





> 5\. Compared to ebview, GD-ng returns a fraction of the search results. For instance, searching 広辞苑 for 赤 causes ebview to return several hundred entries with that character in their headword, whereas GD-ng returns about 40

there should be more than 100 items.
![image](https://github.com/xiaoyifang/goldendict-ng/assets/105986/9792b619-3a8c-4ba1-9a62-8004cb952cfd)

**Comment by twwn** [comment]
> should be fixed with #1092

Unfortunately that patch causes most Epwing / dictionary directories to no longer be loaded at all.

>can you upload one of the dictionary for debug

They're on the web, for instance linked on that "cornucopia" site. The paths in my comment mirror those in the download location offered there.

> This is designed . Epwing is not a dictionary format,but an ebook format , some headword is indeed too long.

Okay. The weird part still is (see my edit) that the headword seems several entries garbled together. The other "too long" reports show the same issue.

> there should be more than 100 items.

I can't check atm (since 広辞苑 is one of the no longer loaded dictionaries). One thing to look out for though, the test needs to be done with a dictionary group that has 広辞苑 as sole member, otherwise the suggestions from the group's other dictionaries ofc get merged together.

**Comment by xiaoyifang** [comment]
> 1. GD-ng seems to search for non-existent EPWING folders. Among others, that yields

https://github.com/xiaoyifang/goldendict-ng/pull/1093  combine the previous patch into this PR. \
folder is case senstive in Linux

**Comment by xiaoyifang** [comment]
> 3\. There also some failures with existing folders (which GD reports as lower-case when they are actually in caps) that cause other errors:
> 
> ```
> Epwing: Building the index for dictionary in directory /dict/Kokugo_misc/Tegami/kokugo [exists, but as KOKUGO]
> "Epwing dictionary initializing failed: /dict/Kokugo_misc/Tegami/kokugo, error: EB library report error EB \"eb_forward_text\" function error: EB_ERR_UNEXP_TEXT (unexpected format in a text file) on page 14979, offset 1626\n"
> ```

This seems a bug in gd.

**Comment by xiaoyifang** [comment]
> [This Goldendict bug](https://github.com/goldendict/goldendict/issues/948) also persists unchanged.

This is not an error ,it just a warning .GoldenDict will try to index the ebook using the headword found.   If it can not found the headword ,it will invoke libeb's built-in methods  eb_search_work & eb_search_exact_word to query for the results.    If there is no results found ,this warning will be print.

**Comment by xiaoyifang** [comment]
I think this has been fixed .  I close it for now .  Feel free to reopen it if new issue arises.
